### ISSUE TITLE
Make it easy for init scripts to dynamically configure gradle.properties files values

### ISSUE THREAD

**Issue by steffenyount** [gradle/gradle#18353]
Provide 1st class support for using init scripts to dynamically configuring the default property values set in the `gradle.properties` file's `installation`, `root`, and `user` hierarchy.

### Expected Behavior
I would like to be able to dynamically set default property values for my custom Gradle distribution and allow those values to be overridden by a project's `gradle.properties` file, a user's `gradle.properties` file, and/or `-P` command line options.

It would be nice if a simple interface were exposed in the init scripts that allowed me to configure property values at any of the `gradle.properties` files search path levels: `installation`/`root`/`user`

I could imagine this being implemented with a new public GradleProperties interface like:
```
public interface GradleProperties {
    MapProperty<String, String> getInstallationProperties();
    MapProperty<String, String> getRootProperties();
    MapProperty<String, String> getUserProperties();
}
```
made accessible to the init scripts via the Gradle interface like:
```
public interface Gradle extends PluginAware {
...
    GradleProperties getGradleProperties();
    void gradleProperties(Action<? super GradleProperties> action);
...
}
```
that way I could write a simple init script to do the job like:

ci.init.gradle
```
gradleProperties {
    installationProperties.put('artifactoryBuildPublishRepo', 'initial value');
}
```

### Current Behavior
What's currently available is quite limited:

1) I can use the ext mechanism to push values into the rootProject with something like the following, but this is all-or-nothing and it doesn't provide the nuance necessary that would allow my default values to be overridden at different levels in the installation/root/user search path.

ci.init.gradle.kts
```
rootProject {
    if (!hasProperty("artifactoryBuildPublishRepo")) {
        val artifactoryBuildPublishRepo by extra("initial value")
    }
}
```

2) I can inject the internal `GradleProperties` object to get a handle on it, but then calling `gradleProperties.mergeProperties()` on it has no effect since the result is immutable, and thus the existing GradleProperties object is not very useful without a bunch of Java reflection hackhackery.

### Context
I'm working on putting together some custom Gradle distributions targeting Gradle 7+ (like what's described here: https://docs.gradle.org/current/userguide/organizing_gradle_projects.html#sec:custom_gradle_distribution ).

For this custom Gradle distribution, I'm trying to augment the static default property values which I have configured via a gradle.properties file and included in the root of my distribution with some dynamically generated "default" property values that I configure via an `init.d/ci.init.gradle` script which I'm also including in the distribution.

This has not been an easy task. The init scripts are not doing basic things that I've grown to expect from working in the settings.gradle and build.gradle scripts. It feels like everywhere I turn I need to dig through the Gradle source files to figure out another "work-around" to get things going again.

I would like this experience to be better, so I'll be filing 5 tickets with hope that we can to improve this situation.

This is ticket 4 of 5

**Comment by ljacomet** [comment]
Can you please expand a bit on why solution 1 does not allow you to do the necessary overrides?

**Comment by steffenyount** [comment]
so here's the capabilities/limitations of solution 1 that I've been able to suss out so far:  

The ExtraPropertiesExtension (a.k.a `ext` in groovy scripts) properties take priority over `gradle.properties` files' properties when set in the same script, but the `gradle.properties` files' properties will take priority when ExtraPropertiesExtension properties are being inherited from a parent project.

ExtraPropertiesExtension properties are not shared between the 3 different types of script objects Gradle, Settings, and Project.

Implications:

If I'm only trying to configure a new default property value for a custom Gradle distribution like the `artifactoryBuildPublishRepo` property from my example, then solution 1 is sufficient, though awkward, as it does require me to set the value 2 times, once for the Settings object and once for the RootProject's Project object for the default property value to be universally accessible wherever a gradle.properties file property would have been accessible.

the existing complex solution:
```
rootProject {
    if (!hasProperty('artifactoryBuildPublishRepo')) {
        ext.artifactoryBuildPublishRepo = 'repo_name'
    }
}
beforeSettings {
    if (!hasProperty('artifactoryBuildPublishRepo')) {
        ext.artifactoryBuildPublishRepo = 'repo_name'
    }
}
```
vs the proposed simpler solution:
```
gradleProperties {
    installationProperties.put('artifactoryBuildPublishRepo', 'repo_name');
}
```

If I want to do something more sophisticated, like add an init script into my `USER_HOME/.gradle/init.d/` to check and then *maybe* replace a default property value that was already defined. E.g: I want it to replace the `repoBaseUrl` property value to point to my regional repo mirror instead of the default repo, but then only apply this replacement when the existing `repoBaseUrl` property actually matches the url of my mirror repo's upstream repo. This then allows my user init script to avoid accidental property name collisions when building 3rd party projects that happened to also use a property named `repoBaseUrl`.

the existing complex solution:
```
```

**Comment by steffenyount** [comment]
so here's the capabilities/limitations of solution 1 that I've been able to suss out so far:  

The ExtraPropertiesExtension (a.k.a `ext` in groovy scripts) properties take priority over `gradle.properties` files' properties when set in the same script, but the `gradle.properties` files' properties will take priority when ExtraPropertiesExtension properties are being inherited from a parent project.

ExtraPropertiesExtension properties are not shared between the 3 different types of script objects Gradle, Settings, and Project.

These rules ^^^ of precedence were not something that I found in the docs, but rather I had to figure them out with experimentation.

 In contrast the `gradle.properties` file's rules of precedence are well defined and can be found in more than one part of the documentation:
`override properties` (cmd line options etc) -> `user properties` -> `project properties` -> `root properties` -> `installation properties`

Implications:

1) If I'm only trying to configure a new default property value for a custom Gradle distribution like the `artifactoryBuildPublishRepo` property from my example, then solution 1 is sufficient, though awkward, as it does require me to set the value 2 times, once for the Settings object and once for the RootProject's Project object for the default property value to be universally accessible wherever a gradle.properties file property would have been accessible.

the existing complex init script solution:
```
rootProject {
    if (!hasProperty('artifactoryBuildPublishRepo')) {
        ext.artifactoryBuildPublishRepo = 'repo_name'
    }
}
beforeSettings {
    if (!hasProperty('artifactoryBuildPublishRepo')) {
        ext.artifactoryBuildPublishRepo = 'repo_name'
    }
}
```
vs the proposed simpler init script solution:
```
gradleProperties {
    installationProperties.put('artifactoryBuildPublishRepo', 'repo_name');
}
```

2) If I want to do something more sophisticated, like add an init script into my `USER_HOME/.gradle/init.d/` to check and then *maybe* replace a default property value that was already defined. Then the ExtraPropertiesExtension gets even more awkward, since now I need to set the `ext` property value on *all* of the Project objects and the Settings object. E.g: I want it to replace the `repoBaseUrl` property value to point to my regional repo mirror instead of the default repo, but then only apply this replacement when the existing `repoBaseUrl` property actually matches the url of my mirror repo's upstream repo. This then allows my user init script to avoid accidental property name collisions when building 3rd party projects that happened to also use a property named `repoBaseUrl`.

the existing complex init script solution:
```
rootProject {
    allProjects {
        if (!hasProperty('artifactoryBuildPublishRepo')) {
            ext.artifactoryBuildPublishRepo = 'repo_name'
        }
    } 
}
beforeSettings {
    if (!hasProperty('artifactoryBuildPublishRepo')) {
        ext.artifactoryBuildPublishRepo = 'repo_name'
    }
}
```
vs the proposed simpler init script solution:
```
gradleProperties {
    if (hasProperty('repoBaseUrl') && 'recognized_repo_base_url'.equals(repoBaseUrl)) {
            userProperties.put('repoBaseUrl', 'mirror_repo_base_url');
    }
}
```
